src/otaclient/create_standby/resume_ota.py

```diff
@@ -149,3 +190,142 @@ def resume_ota(self) -> None:
+            status_reporter_t.join()
+
+
+class ResourceStreamer(_ResourceOperatorBase):
+    """Scan and verify OTA resource folder at the source slot, and stream with verifying it
+    to the destination slot.
+    """
+
+    def __init__(
+        self,
+        *,
+        all_resource_digests: ResourcesDigestWithSize,
+        src_resource_dir: Path,
+        dst_resource_dir: Path,
+        status_report_queue: Queue[StatusReport],
+        session_id: str,
+    ) -> None:
+        self._status_report_queue = status_report_queue
+        self.session_id = session_id
+        self._src_resource_dir = src_resource_dir
+        self._dst_resource_dir = dst_resource_dir
+        self._all_resource_digests = all_resource_digests
+
+        self._se = threading.Semaphore(cfg.MAX_CONCURRENT_PROCESS_FILE_TASKS)
+        self._thread_local = threading.local()
+        self._internal_que: Queue[int | None] = Queue()
+
+    def _process_resource_at_thread(
+        self,
+        expected_digest: bytes,
+        expected_digest_hex: str,
+        *,
+        src_dir_fd: int,
+        dst_dir_fd: int,
+    ) -> None:
+        tmp_dst_fname = _gen_tmp_fname()
+        try:
+            hash_f, file_size = sha256(), 0
+            buffer, bufferview = (
+                self._thread_local.buffer,
+                self._thread_local.bufferview,
+            )
+
+            src_fd = os.open(expected_digest_hex, os.O_RDONLY, dir_fd=src_dir_fd)
+            tmp_dst_fd = os.open(
+                tmp_dst_fname,
+                os.O_WRONLY | os.O_CREAT | os.O_TRUNC,
+                dir_fd=dst_dir_fd,
+            )
+
+            with open(src_fd, "rb") as src, open(tmp_dst_fd, "wb") as tmp_dst:
+                os.posix_fadvise(src_fd, 0, 0, os.POSIX_FADV_NOREUSE)
+                os.posix_fadvise(src_fd, 0, 0, os.POSIX_FADV_SEQUENTIAL)
+                os.posix_fadvise(tmp_dst_fd, 0, 0, os.POSIX_FADV_NOREUSE)
+                os.posix_fadvise(tmp_dst_fd, 0, 0, os.POSIX_FADV_SEQUENTIAL)
+                while read_size := src.readinto(buffer):
+                    file_size += read_size
+                    tmp_dst.write(bufferview[:read_size])
+                    hash_f.update(bufferview[:read_size])
+                os.posix_fadvise(src_fd, 0, 0, os.POSIX_FADV_DONTNEED)
+                os.posix_fadvise(tmp_dst_fd, 0, 0, os.POSIX_FADV_DONTNEED)
+
+            calculated_digest = hash_f.digest()
+            if calculated_digest != expected_digest:
+                self._remove_entry(tmp_dst_fname, dir_fd=dst_dir_fd)
+                return
+
+            try:
+                self._all_resource_digests.pop(calculated_digest)
+                # NOTE: both src and dst are under the same folder now
+                os.replace(
+                    tmp_dst_fname,
+                    expected_digest_hex,
+                    src_dir_fd=dst_dir_fd,
+                    dst_dir_fd=dst_dir_fd,
+                )
```

Comment:
do we need `flush` and `fsync` for the target file?

Reply:
os.replace is calling rename sys call, I think kernel will take care of it.